src/tpm2/NVMarshal.c

```diff
@@ -2026,7 +2026,145 @@ tpmHashStateSHA512_Unmarshal(SHA512_CTX *data, BYTE **buffer, INT32 *size,
     return rc;
 }
 #endif
+#if ALG_SM3_256
+
+#define HASH_STATE_SM3_256_MAGIC 0x10854a09
+#define HASH_STATE_SM3_256_VERSION 2
+
+#include "Sm3Helper_fp.h"
+static inline UINT16
+SM3_WORD_Marshal(SM3_WORD *data, BYTE **buffer, INT32 *size)
+{
+    return UINT32_Marshal(data, buffer, size);
+}
+
+static inline UINT16
+SM3_WORD_Unmarshal(SM3_WORD *data, BYTE **buffer, INT32 *size)
+{
+    return UINT32_Unmarshal(data, buffer, size);
+}
+static UINT16
+tpmHashStateSM3_256_Marshal(tpmHashStateSM3_256_t *data, BYTE **buffer, INT32 *size,
+                           UINT16 hashAlg)
+{
+    UINT16 written = 0;
+    UINT16 array_size;
+    SM3_CTX *sm3_ctx = NULL;
+    BLOCK_SKIP_INIT;
+
+    sm3_ctx = EVP_MD_CTX_md_data(*data);
+    written = NV_HEADER_Marshal(buffer, size,
+                                HASH_STATE_SM3_256_VERSION,
+                                HASH_STATE_SM3_256_MAGIC, 1);
+    written += SM3_WORD_Marshal(&sm3_ctx->A, buffer, size);
+    written += SM3_WORD_Marshal(&sm3_ctx->B, buffer, size);
+    written += SM3_WORD_Marshal(&sm3_ctx->C, buffer, size);
+    written += SM3_WORD_Marshal(&sm3_ctx->D, buffer, size);
+    written += SM3_WORD_Marshal(&sm3_ctx->E, buffer, size);
+    written += SM3_WORD_Marshal(&sm3_ctx->F, buffer, size);
+    written += SM3_WORD_Marshal(&sm3_ctx->G, buffer, size);
+    written += SM3_WORD_Marshal(&sm3_ctx->H, buffer, size);
+    written += SM3_WORD_Marshal(&sm3_ctx->Nl, buffer, size);
+    written += SM3_WORD_Marshal(&sm3_ctx->Nh, buffer, size);
+    /* data must be written as array */
+    array_size = sizeof(sm3_ctx->data);
+    written += UINT16_Marshal(&array_size, buffer, size);
+    written += Array_Marshal((BYTE *)&sm3_ctx->data[0], array_size, buffer, size);
+    written += UINT32_Marshal(&sm3_ctx->num, buffer, size);
+    written += BLOCK_SKIP_WRITE_PUSH(TRUE, buffer, size);
+    /* future versions append below this line */
 
+    BLOCK_SKIP_WRITE_POP(size);
+
+    BLOCK_SKIP_WRITE_CHECK;
+
+    return written;
+}
+static UINT16
+tpmHashStateSM3_256_Unmarshal(tpmHashStateSM3_256_t *data, BYTE **buffer, INT32 *size,
+                             UINT16 hashAlg)
+{
+    UINT16 rc = TPM_RC_SUCCESS;
+    UINT16 array_size;
+    NV_HEADER hdr;
+    SM3_CTX *sm3_ctx = NULL;
+    
+    (*data) = EVP_MD_CTX_new();
```

Comment:
I think we shouldn't create a context here but fill the data just like we fill it with sha1 etc.

Reply:
The risk with sm3 is that its data structure has always been opaque it seems. That leaves OpenSSL the opportunity to shuffle things around. They may not do that but the risk is there. In contrast to that we have the SHAT_CTX, SHA256_CTX, etc. that are available via sha.h, so there's less of a risk with them on that level.

Reply:
I know... but it's not going to be that simple. See my comment over there: https://github.com/stefanberger/libtpms/issues/215#issuecomment-842757244